web/courses/models.py

```diff
@@ -486,46 +309,16 @@ def results_archive(self, user):
         files.append((spreadsheet_filename, spreadsheet_contents))
         return archive_name, files
 
-    def student_statistics(self):
-        students = self.course.observed_students()
-        student_count = len(students)
-
-        attempts = Attempt.objects.filter(
-            part__problem__problem_set=self, user__in=students
-        ).values("part_id", "valid")
-        attempts_dict = {}
-        for attempt in attempts:
-            part_id = attempt["part_id"]
-            if part_id in attempts_dict:
-                attempts_dict[part_id]["valid_count"] += 1 if attempt["valid"] else 0
-                attempts_dict[part_id]["submitted_count"] += 1
-            else:
-                attempts_dict[part_id] = {
-                    "submitted_count": 1,
-                    "valid_count": 1 if attempt["valid"] else 0,
-                }
-
+    def outcomes_statistics(self, outcomes):
         statistics = []
-        for problem in self.problems.all():
+        for problem in self.problems.prefetch_related("parts"):
```

Comment:
Ne spoznam se najbolje na uporabo te funkcije; je kakšen poseben razlog, da se tu (in v `problem_set_detail`) uporablja samo `X.prefetch_related(...)` namesto `X.all().prefetch_related(...)`, kakor v ostalih primerih v kodi?

Reply:
`X` (v tem primeru `self.problems`, lahko pa tudi npr. `Part.objects`) je `manager`, torej nekaj za delo z množico primerkov modela. Npr. z njim se lahko dodaja/briše primerke, glavna uporaba pa so poizvedbe. Vsaka poizvedba vrne `QuerySet`, čez katerega se lahko iterira. Če bi naredili samo `for ... in X`, bi dobili napako. Običajno se uporabi `for ... in X.all()`, kjer je `.all` metoda, ki vrne `QuerySet` vseh modelov. Zaradi lažjega veriženja lahko večina metod za poizvedbe (`filter`, `exclude`, `prefetch_related`, ...) sprejme tako `Manager` kot `QuerySet`, vedno pa vrne `QuerySet`. Tako da v primeru, ko imamo `X.prefetch_related()`, ni treba pisati `X.all()`, ampak tudi nič ne škodi.

Reply:
Aha, hvala.